### PR TITLE
Add query for Tavern, Dwelling and Market visits

### PR DESCRIPTION
**Pull request by IvanSavenko**
Adventure map Tavern and all markets now generate queries for duration of the visit until client closes corresponding window. Taverns / markets in towns do not generate queries

- Fixes broken visitation to these buildings from recent PR.
- Allows proper validation of requests from players by server

### CODE REVIEW

File: server/queries/MapQueries.cpp [hunk 1]
```diff
@@ -157,6 +157,50 @@ CBlockingDialogQuery::CBlockingDialogQuery(CGameHandler * owner, const BlockingD
...
+			return false;
+
+		if(dynamic_ptr_cast<AssembleArtifacts>(pack))
+			return false;
+
+		if(dynamic_ptr_cast<EraseArtifactByClient>(pack))
+			return false;
+
+		if(dynamic_ptr_cast<TradeOnMarketplace>(pack) != nullptr)
+			return false;
```
Comment (Nordsoft91) [inline server/queries/MapQueries.cpp:198]:
```suggestion
		if(dynamic_ptr_cast<HireHero>(pack))
			return false;
	}

	if (mode == EOpenWindowMode::UNIVERSITY_WINDOW)
	{
		if(dynamic_ptr_cast<TradeOnMarketplace>(pack))
			return false;
	}

	if (mode == EOpenWindowMode::MARKET_WINDOW)
	{
		if(dynamic_ptr_cast<ExchangeArtifacts>(pack))
			return false;

		if(dynamic_ptr_cast<AssembleArtifacts>(pack))
			return false;

		if(dynamic_ptr_cast<EraseArtifactByClient>(pack))
			return false;

		if(dynamic_ptr_cast<TradeOnMarketplace>(pack))
			return false;
```

Comment (Nordsoft91) [reply server/queries/MapQueries.cpp:198]: Can we inherit all those packs from one empty class to have only single cast check? Thus, we can simplify adding of new packs in the future

File: AI/Nullkiller/AIGateway.cpp [hunk 2]
```diff
@@ -154,10 +154,13 @@ void AIGateway::artifactAssembled(const ArtifactLocation & al)
...
 }
 
-void AIGateway::showTavernWindow(const CGObjectInstance * townOrTavern)
+void AIGateway::showTavernWindow(const CGObjectInstance * object, const CGHeroInstance * visitor, QueryID queryID)
 {
 	LOG_TRACE(logAi);
 	NET_EVENT_HANDLER;
+
+	status.addQuery(queryID, "TavernWindow");
+	requestActionASAP([=](){ answerQuery(queryID, 0); });
```
Comment (Nordsoft91) [inline AI/Nullkiller/AIGateway.cpp:163]: I was thinking NKAI uses advmap taverns to hire heroes too

File: server/queries/MapQueries.cpp (hunk 1 above)
Comment (IvanSavenko) [reply server/queries/MapQueries.cpp:208]: Probably not this time.
We also have similar (and even longer) set of checks for exchange dialog.
We also may want to examine content of netpack to allow only packs that affect objects involved in query so simple empty class won't work.

File: AI/Nullkiller/AIGateway.cpp (hunk 2 above)
Comment (IvanSavenko) [reply AI/Nullkiller/AIGateway.cpp:163]: Does not looks like it. From what I see NKAI can only recruit heroes from towns.

Comment (nullkiller) [reply AI/Nullkiller/AIGateway.cpp:163]: Yep, I do not remember such functionality so expect that it can not

**Pull request merged**